Docs: Plugin doc review for chunk 1-A (#67070)
grafana/grafana: docs/sources/developers/plugins/build-a-streaming-data-source-plugin.md

# Build a streaming data source plugin

In Grafana, you can set your dashboards to automatically refresh at a certain interval, no matter what data source you use. Unfortunately, this means that your queries are requesting all the data to be sent again, regardless of whether the data has actually changed. The solution to this issue is to add streaming to the plugin to update your dashboard only when new data becomes available.

## Before you begin

This guide assumes that you're already familiar with how to [Build a data source plugin](/tutorials/build-a-data-source-plugin/).

Grafana uses [RxJS](https://rxjs.dev/) to continuously send data from a data source to a panel visualization.

> **Note:** To learn more about RxJs, refer to the [RxJS documentation](https://rxjs.dev/guide/overview).

## Add streaming to your data source

> Enable streaming for your data source plugin to update your dashboard as soon as new data becomes available.

Reviewer: ```suggestion
Enable streaming for your data source plugin to update your dashboard when new data becomes available.
```

> # Build a streaming data source plugin

Reviewer 2: should this be a sub page of Create a Grafana plugin?

Author: Yes, it will move there in phase 2.

> In Grafana, you can set your dashboards to automatically refresh at a certain interval, no matter what data source you use. Unfortunately, this means that your queries are requesting all the data to be sent again, regardless of whether the data has actually changed. The solution to this issue is to add _streaming_ to the plugin.

Reviewer 2: I would extend on why it is the solution here rather than below. My suggestion would be "to update your dashboard only when new data becomes available."

> In Grafana, you can set your dashboards to automatically refresh at a certain interval, no matter what data source you use. Unfortunately, this means that your queries are requesting all the data to be sent again, regardless of whether the data has actually changed. The solution to this issue is to add _streaming_ to the plugin.

Author: Fixed in latest commit.

> This guide explains how to build a streaming data source plugin.

Reviewer: Do you need this sentence? Your paragraph below this one provides a nice intro.

Author: Done, will check fix in soon

Author: Checked in fix, available for review.

In Grafana, you can set your dashboards to automatically refresh at a certain interval, no matter what data source you use. Unfortunately, this means that your queries are requesting all the data to be sent again, regardless of whether the data has actually changed. The solution to this issue is to add _streaming_ to the plugin.

## Before you begin

This guide assumes that you're already familiar with how to [Build a data source plugin](/tutorials/build-a-data-source-plugin/).

> Grafana uses [RxJS](https://rxjs.dev/) to continuously send data from a data source to a panel visualization. There's a lot more to RxJS than what's covered in this guide. 

Reviewer: Since the note below points to the RxJS documentation, it's redundant to say that there's more than what's covered in this guide since that is shown by the doc link. 
```suggestion
Grafana uses [RxJS](https://rxjs.dev/) to continuously send data from a data source to a panel visualization. 
```

Author: Fixed, will check in soon

Author: Checked in fix, available for review.

   };
   ```

>    > **Note:** In practice, you'd call `subscriber.next` as soon as you receive new data from a websocket or a message bus. In the example above, data is being received every 500 milliseconds.

Reviewer 2: I found it strange that we're showing this... what do folks think of instead linking to and referencing our sample https://github.com/grafana/grafana-plugin-examples/tree/main/examples/datasource-streaming-websocket?

Author: my 2 cents: I think we want at least one good, detailed example in each how-to/tutorial in Create a plugin; and we should add links to relevant samples in grafana-plugin-examples. That said, if this isn't a strong example to show for whatever reason, let's remove it until a better example is created.

Author: I'd like to postpone deleting the example until we get to phase 3.

Author: I have created https://github.com/grafana/grafana/issues/67566 to track.

   };
   ```

   > **Note:** In practice, you'd call `subscriber.next` as soon as you receive new data from a websocket or a message bus. In the example above, data is being received every 500 milliseconds.

> ### Example code for final `query` method

Reviewer 3: I don't know if this is linked somewhere else in this doc but we should point to https://github.com/grafana/grafana-plugin-examples/tree/main/examples/datasource-streaming-websocket for a full example of a streaming plugin

Author: Good idea. Will check in fix soon.

Author: Checked in fix, available for review.
